src/Pos/Communication/Limits.hs

```diff
@@ -0,0 +1,455 @@
+updateVoteNumLimit = round $ 1 / coinPortionToDouble Const.genesisUpdateVoteThd
+
+-- | Upper bound on size of `upData` in `UpdateProposal`.
+upDataNumLimit :: Int
+upDataNumLimit = 1
+
+-- | Upper bound on size of `ApplicationName`
+appNameLenLimit :: Int
+appNameLenLimit = 30
+
+vectorOf :: IsList l => Int -> Limit (Item l) -> Limit l
+vectorOf k (Limit x) =
+    Limit $ encodedListLength + x * (fromIntegral k)
+  where
+    -- should be enough for most reasonable cases
+    encodedListLength = 20
+
+vector :: (IsList l, MessageLimitedPure (Item l)) => Int -> Limit l
+vector k = vectorOf k msgLenLimit
+
+multiMap
+    :: (IsList l, Item l ~ (k, l0), IsList l0,
+        MessageLimitedPure k, MessageLimitedPure (Item l0))
+    => Int -> Limit l
+multiMap k =
+    -- max message length is reached when each key has single value
+    vectorOf k $ (,) <$> msgLenLimit <*> vector 1
+
+coerce :: Limit a -> Limit b
+coerce (Limit x) = Limit x
+
+-- | Specifies type of limit on incoming message size.
+-- Useful when the type has several limits and choice depends on constructor.
+class Limiter l where
+    limitGet :: l -> Get a -> Get a
+    addLimit :: Byte -> l -> l
+
+instance Limiter (Limit t) where
+    limitGet (Limit l) = Bi.limitGet $ fromIntegral l
+    addLimit a = mappend (Limit a)
+
+-- | Bounds `InvOrData`.
+instance Limiter l => Limiter (Limit t, l) where
+    limitGet (invLimit, dataLimits) parser = do
+        lookAhead getWord8 >>= \case
+            0   -> limitGet invLimit parser
+            1   -> limitGet dataLimits parser
+            tag -> fail ("get@InvOrData: invalid tag: " ++ show tag)
+
+    addLimit a (l1, l2) = (a `addLimit` l1, a `addLimit` l2)
+
+-- | Bounds `DataMsg`.
+-- Limit depends on value of first byte, which should be in range @0..3@.
+instance Limiter (Limit t, Limit t, Limit t, Limit t) where
+    limitGet limits parser = do
+        tag <- fromIntegral <$> lookAhead getWord8
+        case (limits ^.. each) ^? ix tag of
+            Nothing -> fail ("get@DataMsg: invalid tag: " ++ show tag)
+            Just limit -> limitGet limit parser
+
+    addLimit a = both %~ addLimit a
+
+-- | Specifies limit on message length.
+-- Deserialization would fail if incoming data size exceeded this limit.
+-- At serialisation stage message size is __not__ checked.
+class Limiter (LimitType a) => MessageLimited a where
+    type LimitType a :: *
+    getMsgLenLimit :: DB.MonadDBLimits m => Proxy a -> m (LimitType a)
+
+instance MessageLimited (MsgBlock ssc) where
+    type LimitType (MsgBlock ssc) = Limit (MsgBlock ssc)
+    getMsgLenLimit _ = Limit <$> DB.getMaxBlockSize
+
+instance MessageLimited MsgGetBlocks where
+    type LimitType MsgGetBlocks = Limit MsgGetBlocks
+    getMsgLenLimit _ = return msgLenLimit
+
+instance MessageLimited MsgGetHeaders where
+    type LimitType MsgGetHeaders = Limit MsgGetHeaders
+    getMsgLenLimit _ = return $
+        MsgGetHeaders <$> vector Const.genesisMaxGetHeadersNum <*> msgLenLimit
```

Comment:
Regarding better limitation than just randomly set 100: well, we can compute it using current block size and slot duration. For instance, if slot duration is 5 sec (who knows, maybe in 20 years it will be even less, probably there will be no slots in a year, but let's temporary forget about it :D) it probably means that we can and should process more blocks (it's a good indicator that computers' capabilities were improved).
I think we shouldn't put much effort into such things though.